cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java

```diff
@@ -243,8 +243,24 @@ private Object convertString(Schema schema, String strVal) throws UnexpectedForm
       if (field == null) {
         throw new UnexpectedFormatException("field " + fieldName + " is not in the schema.");
       }
-      if (!field.getSchema().isNullable() && val == null) {
-        throw new UnexpectedFormatException("field " + fieldName + " cannot be set to a null value.");
+      Schema fieldSchema = field.getSchema();
+      if (val == null) {
+        if (fieldSchema.getType() == Schema.Type.NULL) {
+          return field;
+        }
+        if (fieldSchema.getType() != Schema.Type.UNION) {
+          throw new UnexpectedFormatException("field " + fieldName + " cannot be set to a null value.");
+        }
+        boolean ok = false;
+        for (Schema unionSchema : fieldSchema.getUnionSchemas()) {
+          if (unionSchema.getType() == Schema.Type.NULL) {
+            ok = true;
```

Comment:
You can just have `return null` here without the usage of `ok` variable and throw exception unconditionally if the for loop ever exit.